crates/matrix-sdk-common/src/wasm_tracing.rs

```diff
@@ -0,0 +1,129 @@
+// Copyright 2023 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::{fmt, fmt::Write};
+
+use tracing::{
+    field::{Field, Visit},
+    metadata::LevelFilter,
+    Event, Level, Metadata, Subscriber,
+};
+use tracing_subscriber::layer::Context;
+
+///! Utilities for `tracing` in wasm environments
+
+/// An implementation of `tracing_subscriber::layer::Layer` which directs all
+/// events to the JS console
+#[derive(Debug)]
+pub struct Layer {
+    min_level: Level,
+    enabled: bool,
+}
+
+impl Layer {
+    pub fn new<L>(min_level: L) -> Self
+    where
+        L: Into<Level>,
+    {
+        Self { min_level: min_level.into(), enabled: true }
+    }
+
+    pub fn turn_on(&mut self) {
+        self.enabled = true;
+    }
+
+    pub fn turn_off(&mut self) {
+        self.enabled = false;
+    }
+
+    pub fn set_min_level(&mut self, min_level: Level) {
+        self.min_level = min_level;
+    }
```

Comment:
This seems like a reimplementation of [`LevelFilter`](https://docs.rs/tracing-subscriber/latest/tracing_subscriber/filter/struct.LevelFilter.html) which can be applied to a layer using [`with_filter`](https://docs.rs/tracing-subscriber/latest/tracing_subscriber/layer/trait.Layer.html#method.with_filter) or applied to a `tracing_subscriber::Registry` before the layer that consumes the logs.

I think it would make sense to use that layer instead.

Reply:
oh, I figured it out. Will update later